Robot2018/test/org/usfirst/frc/team199/Robot2018/AutoLiftTest.java

```diff
@@ -0,0 +1,179 @@
+
+class AutoLiftTest {
+	
+	final double liftkP = 0.01;
+	final double liftkI = 0.0001;
+	final double liftkD = 0.002;
+	private final double targetOne = 24;
+	private final double targetTwo = 60;
+	private final double targetThree = 45;
+	private int iterations;
+	private final double tolerance = 1;
+	
+	@BeforeEach
+	void setUp() {
+		// Since VelocityPIDController extends PIDController and PIDController calls
+		// static methods in wpilib that only work on robot,
+		// we setup these mocks to allow the tests to run off robot.
+		iterations = 0;
+		HardwareTimer tim = mock(HardwareTimer.class);
+		Timer.Interface timerInstance = mock(Timer.Interface.class);
+		when(tim.newTimer()).thenReturn(timerInstance);
+		when(timerInstance.get()).thenAnswer(
+			     new Answer() {
+			         public Object answer(InvocationOnMock invocation) {
+			             return getTime();
+			         }
+			 });
+		
+		Timer.SetImplementation(tim);
+		HLUsageReporting.Interface usageReporter = mock(HLUsageReporting.Interface.class);
+		HLUsageReporting.SetImplementation(usageReporter);
+	}
+	
+	private double getTime() {
+		iterations++;
+		return iterations * 0.02;
+	}
+	
+	class PIDSim{
+		private double totalDist = 0;
+		private final double distPerPulse = 0.01;
+		// 1/pulses per rev, 1/99?
+		private final double circum = Math.PI * 6;
+		public PIDSim() {
+			totalDist = 0;
+		}
+		
+		public double pidGet() {
+			return totalDist;
+		}
+		
+		public void pidWrite(double output) {
+			//256 rpm
+			totalDist += (output * 256) * distPerPulse * circum;
+		}
+	}
+	
+	class PIDSimController extends PIDController{
+		
+		public PIDSimController(double p, double i, double d, PIDSource src, PIDOutput out) {
+			super(p,i,d,src,out);
+		}
+		
+		public void calc() {
+			this.calculate();
+		}
+		
+	}
+	
+	class PIDSimSource implements PIDSource{
+		private PIDSim sim;
+		private PIDSourceType type;
+		public PIDSimSource(PIDSim sim) {
+			this.sim = sim;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return type;
+		}
+
+		@Override
+		public double pidGet() {
+			return sim.pidGet();
+		}
+		
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			// TODO Auto-generated method stub
+			type = pidSource;
+		}
+	}
+	
+	class PIDSimOutput implements PIDOutput{
+		private PIDSim sim;
+		public PIDSimOutput(PIDSim sim) {
+			this.sim = sim;
+		}
+		@Override
+		public void pidWrite(double output) {
+			sim.pidWrite(output);
+		}
+	}
+
+	@Test
+	void test24() {
+		PIDSim sim = new PIDSim();
+		PIDSimSource src = new PIDSimSource(sim);
+		src.setPIDSourceType(PIDSourceType.kDisplacement);
+		PIDSimOutput out = new PIDSimOutput(sim);
+		PIDSimController liftController = new PIDSimController(liftkP, liftkI, liftkD, src, out);
+		liftController.setSetpoint(targetOne);
+		liftController.setOutputRange(-1.0, 1.0);
+		liftController.enable();
+		//1 second?
+		for(int i = 0; i < 50; i++) {
+			liftController.calc();
+		}
```

Comment:
I believe the PIDController's thread will also be calling calculate() periodically. This will be in addition to your calls via calc(). The combination could cause the test to be unreliable.

I recommend overriding PIDController.calculate() so that it only calls super.calculate() if Thread.currentThread().getId() matches the thread id being used by the simulator.

Reply:
I assumed that would be the case as well, but after looking at the source code it doesn't seem to work that way. All enable() does is set an enabled boolean that doesn't necessarily start the thread. I'm positive the current way of calculating is only one thread.

Reply:
The thread (actually a timer that periodically runs a task on a separate thread) is started in the constructor [here](https://github.com/wpilibsuite/allwpilib/blob/738a1c015cfb42076cfda2e2fb796e66bc98db54/wpilibj/src/main/java/edu/wpi/first/wpilibj/PIDController.java#L174). Still positive? :)

Reply:
The reason I was positive was because I tested the loop by printing the current position rather than calculating after enabling the controller, it was zero all throughout. But the print lines were far faster than the calculating thread so I see now, I'll take your advice and implement calculate that way, thanks.